corpus/codeblock.txt

```diff
@@ -24,13 +24,15 @@ block3:
   (block
     (line
       (codeblock
-        (line))))
+        (code
+          (line)))))
```

Comment:
That's the exact choice I've outlined above: unique names vs. common names and fields. It depends on the exact situation (how often you can reuse a global name). In the case above, it looked like a tradeoff between number of tokens and number of states in the generated parser; unique name led to slightly smaller parser size.

Reply:
You can't _name_ the thing `(block)`, because it _is not_ a `block` -- which contains the terminating `<`.  If you want to reuse the `block` token, you need to rewrite the whole grammar.

Reply:
>  I guess it makes sense for (code) to be named (block). 

But you can't, since `(block)` is already defined at this point (as something different) -- you'll get a conflict in the grammar you need to resolve somehow. And a different name is by far the easiest way to do that.

Reply:
Go ahead, try it.

Reply:
Just tried this, it works fine (no conflicts):

    alias(repeat1(alias($.line_code, $.line)), $.block),

Reply:
You do realize that the alias is to lines (plural) of code?

Reply:
Yes. But clearly I'm missing something or we're miscommunicating, because you seem annoyed.

Reply:
Somewhat, yes, because I don't get what the end goal is here, so it feels like stabbing in the dark, which is frustrating.

(I do admit that my last comment was sheer pedantry, born out of said feeling.)

Reply:
well I started this thread by literally saying "I'm admittedly going in circles". The goal was just to explore options.

Reply:
> Just tried this, it works fine (no conflicts):
> 
> ```
> alias(repeat1(alias($.line_code, $.line)), $.block),
> ```

Interesting, that's what I tried. Looks like switching from `optional` to `choice` avoids the conflict then; that's good. Feel free to rename as you wish before tagging a release so I can downstream the queries.

> Note that (codeline) should be understood as any number of children, e.g.:
> ...
> But this is just academic. If you're ok with `(code (codeline))` , I think I favor that too.

There seems to be indeed some miscommunication. My point is that the contents _must_ be a _single_ block that you pass to the injected parser. Otherwise every single line gets parsed _in isolation_ which a) is wasteful and b) leads to errors with constructs spanning multiple lines (as in my -- admittedly academic -- example above).